source/comp/markv_codec.cpp

```diff
@@ -671,263 +1010,1416 @@ class MarkvDecoder : public MarkvCodecBase {
-        return vstate_.diag(SPV_ERROR_INVALID_BINARY)
-            << "Failed to read literal F64";
-    } else {
-      assert(0);
+  }
+
+  if (spvOpcodeGeneratesType(opcode)) {
+    if (opcode != SpvOpTypeFunction) {
+      multi_mtf_.Insert(kMtfTypeNonFunction, inst_.result_id);
     }
-    spirv_.push_back(static_cast<uint32_t>(word));
-    spirv_.push_back(static_cast<uint32_t>(word >> 32));
   }
-  return SPV_SUCCESS;
-}
 
-spv_result_t MarkvDecoder::DecodeModule(std::vector<uint32_t>* spirv_binary) {
-  const bool header_read_success =
-      reader_.ReadUnencoded(&header_.magic_number) &&
-      reader_.ReadUnencoded(&header_.markv_version) &&
-      reader_.ReadUnencoded(&header_.markv_model) &&
-      reader_.ReadUnencoded(&header_.markv_length_in_bits) &&
-      reader_.ReadUnencoded(&header_.spirv_version) &&
-      reader_.ReadUnencoded(&header_.spirv_generator);
+  const uint32_t descriptor = id_descriptors_.ProcessInstruction(inst_);
 
-  if (!header_read_success)
-    return vstate_.diag(SPV_ERROR_INVALID_BINARY)
-        << "Unable to read MARK-V header";
+  multi_mtf_.Insert(GetMtfIdDescriptor(descriptor), inst_.result_id);
+}
 
-  assert(header_.magic_number == kMarkvMagicNumber);
-  assert(header_.markv_length_in_bits > 0);
+uint64_t MarkvCodecBase::GetRuleBasedMtf(bool* can_forward_declare) {
+  assert(operand_.type != SPV_OPERAND_TYPE_RESULT_ID);
 
-  if (header_.magic_number != kMarkvMagicNumber)
-    return vstate_.diag(SPV_ERROR_INVALID_BINARY)
-        << "MARK-V binary has incorrect magic number";
+  *can_forward_declare = false;
 
-  // TODO(atgoo@github.com): Print version strings.
-  if (header_.markv_version != GetMarkvVersion())
-    return vstate_.diag(SPV_ERROR_INVALID_BINARY)
-        << "MARK-V binary and the codec have different versions";
+  const SpvOp opcode = static_cast<SpvOp>(inst_.opcode);
 
-  spirv_.reserve(header_.markv_length_in_bits / 2); // Heuristic.
-  spirv_.resize(5, 0);
-  spirv_[0] = kSpirvMagicNumber;
-  spirv_[1] = header_.spirv_version;
+  // All operand slots which expect label id.
+  if ((inst_.opcode == SpvOpLoopMerge && operand_index_ <= 1) ||
+      (inst_.opcode == SpvOpSelectionMerge && operand_index_ == 0) ||
+      (inst_.opcode == SpvOpBranch && operand_index_ == 0) ||
+      (inst_.opcode == SpvOpBranchConditional &&
+       (operand_index_ == 1 || operand_index_ == 2 )) ||
+      (inst_.opcode == SpvOpSwitch && operand_index_ > 0)) {
+    *can_forward_declare = true;
+    return kMtfLabel;
+  }
+
+  switch (opcode) {
+    case SpvOpFAdd:
+    case SpvOpFSub:
+    case SpvOpFMul:
+    case SpvOpFDiv:
+    case SpvOpFNegate: {
+      if (operand_index_ == 0)
+        return kMtfTypeFloatScalarOrVector;
+
+      return GetMtfIdOfType(inst_.type_id);
+    }
+
+    case SpvOpISub:
+    case SpvOpIAdd:
+    case SpvOpIMul:
+    case SpvOpSDiv:
+    case SpvOpUDiv:
+    case SpvOpSNegate: {
+      if (operand_index_ == 0)
+        return kMtfTypeIntScalarOrVector;
```

Comment:
Similar to the float case:  For binary operators don't we also want operand index 1?